### PR TITLE
Exclude invalid mirror urls for reposync

### PR DESCRIPTION
**Pull request by zzaimeche**
## What does this PR change?

This commit fixes bsc#1203826. Previously, reposync would treat broken urls in the list of mirrors as valid, causing it to fail. The first commit checks whether urls contain any forbidden characters, and if they do, it discards them. The second commit checks that any page used as a candidate for a mirrorlist is a plaintext or xml page.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage

- [x] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

- [x] **DONE**

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by vzhestkov** [comment]
@zzaimeche thanks for taking care of it.
Just additional suggestion. I've tried before to analyze some of the cases and found that it's possible that the code is trying to get the URLs from the HTML document returned on request for the root of the repo, but in this case it's useless to read the HTML and try to get URLs from it as mirrorlist supposed to be the plain text output if I'm not wrong. Maybe we need to check for the document format returned on request and avoid trying to get list of mirrors in case if it's HTML document?

**Comment by zzaimeche** [comment]
This works fine on my system, but I'm a bit worried that someone might have a strange setup somewhere where they use urls that are forbidden in the RFC (and I left out " as a character, because that's such an edge case and I thought I was more likely to introduce a bug trying to escape it). I still need to write tests, but if users are using weird urls, that won't be caught by tests anyway...

**Comment by zzaimeche** [comment]
> @zzaimeche thanks for taking care of it. Just additional suggestion. I've tried before to analyze some of the cases and found that it's possible that the code is trying to get the URLs from the HTML document returned on request for the root of the repo, but in this case it's useless to read the HTML and try to get URLs from it as mirrorlist supposed to be the plain text output if I'm not wrong. Maybe we need to check for the document format returned on request and avoid trying to get list of mirrors in case if it's HTML document?

From what I understand of the reposync code so far, it goes to the url given by the user, then uses urlgrabber to look through the page at that url for anything that starts with http, ftp, https (and so on), and then when it finds something that matches, it saves *that* to the local textfile (mirrorlist.txt). If there is nothing in the mirrorlist.txt at the end, reposync uses the repourl directly, otherwise it uses urls from that local mirrorlist.txt file. I don't think there's a structural difference between pages with valid lists and the ones without but I might have missed something (though even if people are meant to upload their lists as plain text, I wouldn't be surprised if they're all using different formats anyway).

**Comment by zzaimeche** [comment]
> I think the proper mirror list should be something like this `http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=updates` 

Ah okay, I thought they might be also expecting html pages more like: https://www.debian.org/mirror/list  and making it stricter would break it for users pointing to something like that. (Though I haven't checked many urls yet)

**Comment by vzhestkov** [comment]
@zzaimeche actually I'm not sure about it. And `apt` repos are handled with `deb_src` plugin. I don't remeber if there was any way to process the mirrors.
I think it's better to ask those who have more experience with different kinds of mirrors. Or make this mirror list parsing optional to prevent accidental selection of improper URLs from the page.

**Comment by zzaimeche** [comment]
> @zzaimeche actually I'm not sure about it. And `apt` repos are handled with `deb_src` plugin. I don't remeber if there was any way to process the mirrors. I think it's better to ask those who have more experience with different kinds of mirrors. Or make this mirror list parsing optional to prevent accidental selection of improper URLs from the page.

Okay. I'll try to find out if yum mirrorlists are certain to be plaintext; at the moment I'm a bit worried that users might have their lists of urls up on all sorts of weird pages that just happen to work with the current code, even if they're not supposed to, so fixing it might break things for them in ways that are hard to check in advance. But the current code is really fragile so I'd rather check the document format if I can. Thanks. :)

**Comment by zzaimeche** [comment]
Bit messy and am not thrilled at adding another import, but I couldn't see a way to use urlgrabber for it. I haven't been able to find out yet if only plaintext pages are mirrorlists, so I've kept the commits separate in case that does cause problems.

**Comment by zzaimeche** [comment]
Ugh: https://fedoraproject.org/wiki/Infrastructure/MirrorManager#Metalink_format_mirrorlist

Apparently this is an example: http://ftp5.gwdg.de/pub/opensuse/discontinued/distribution/11.3/iso/ (Content-Type | text/html)

Those pages look really old, so I wondered if it was no longer used, but I found: https://serverfault.com/questions/1021273/all-yum-mirrors-failing-with-errno-1-repomd-xml-does-not-match-metalink which led to an example here: https://mirror.vcu.edu/pub/gnu_linux/epel/7/x86_64/

**Comment by vzhestkov** [comment]
Not sure what you mean. Metalink is the different format which is not supported by reposync, if I'm not wrong.
But mirrorlist from the example `https://mirrors.fedoraproject.org/mirrorlist?repo=fedora-10&arch=x86_64` returns exact plain text list of URLs.

**Comment by zzaimeche** [comment]
> Not sure what you mean. Metalink is the different format which is not supported by reposync, if I'm not wrong. But mirrorlist from the example `https://mirrors.fedoraproject.org/mirrorlist?repo=fedora-10&arch=x86_64` returns exact plain text list of URLs.

Oh, I thought if yum supported it then reposync was expected to support it because it's the yum_src plugin. If we don't have to then that's easier.

**Comment by vzhestkov** [comment]
It could be confusing, but `yum_src` is using `zypper` internally. Reposync plugin does all of these checks for the mirror lists etc. before calling `zypper`.

**Comment by zzaimeche** [comment]
Looks like the metalink support was explictly added: https://github.com/SUSE/spacewalk/commit/4e6f367f1f7b03956c7124e8a72f177a4a9d4ae3 . I think the code's currently structured so that it assumes a list is a mirrorlist, then tries to treat it as a metalink and if that fails it goes back to treating it as a mirrorlist. So I think I could just change my second commit so it checks 'if page plaintext or xml', and that way at least html pages would still be excluded.

**Comment by vzhestkov** [comment]
Oh, sorry I haven't checked deeply to the case when it tries to parse XML. In general it seems that we need to cover only clear text case with mirrorlist. Thanks for pointing :+1:

**Comment by zzaimeche** [comment]
> Oh, sorry I haven't checked deeply to the case when it tries to parse XML. In general it seems that we need to cover only clear text case with mirrorlist. Thanks for pointing +1

Yeah, I think the code is a bit confusingly named/commented because it talks about mirrorlists, but at the start I think it's also getting xml if that's what it finds. So it's called _get_mirror_list but it's also getting the xml stuff if it sees that (unless I'm very confused). I tried to test against a repo with a metalink set as the repourl (this one: https://mirrors.fedoraproject.org/metalink?repo=fedora-18&arch=x86_64 ) to make sure I hadn't broken the xml bit... and it didn't work anyway. =D Not sure if there's a setting I should change somewhere or if that code just never worked or if I just misunderstood it anyway. If it just never worked, I'll change it back to checking for plaintext only. :)

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (vzhestkov) [inline python/spacewalk/satellite_tools/repo_plugins/yum_src.py:531]: It seems in this case it will ignore HTTP_PROXY settings set with web UI (rhn.conf).

File: python/spacewalk/spacewalk-backend.changes
Comment (vzhestkov) [inline python/spacewalk/spacewalk-backend.changes:1]: No need to have 2 distinct entries in the changelog. And no need to set the header for the entry, check with the other PRs for the formatting. For example: https://github.com/uyuni-project/uyuni/pull/6023/files

Comment (zzaimeche) [reply python/spacewalk/spacewalk-backend.changes:1]: ah okay, I followed some linked opensuse docs that said to use the output of `osc vc` so I used that, I'll try to update those too if I remember.

Comment (agraul) [reply python/spacewalk/spacewalk-backend.changes:1]: The "authoritative" source for the changelogs would be https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs (openSUSE changelog guidelines still apply)

**Comment by zzaimeche** [comment]
I've marked it as a wip because I haven't had a chance to check if the proxy code does what I think it does, but I wanted to put it up before I forgot about it at the end of the week!

File: python/spacewalk/spacewalk-backend.changes
Comment (vzhestkov) [inline python/spacewalk/spacewalk-backend.changes:1]: The format of the entry is changing on releasing the package. There is no need to format it this way, just keep it as a one level entry.
Something like:
```
- Use only plain text as mirrorlist (bsc#1203826)
```
as XML is relevant for metalinks and this change is not affecting this part if I'm not wrong.

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (vzhestkov) [inline python/spacewalk/satellite_tools/repo_plugins/yum_src.py:578]: Is it still required as we are not reading HTML as a list of URLs anymore?

**Review by vzhestkov — COMMENTED** [review]
In general looks like it tries to parse the mirrorlist as a plaintext even if it was detected that it's xml. Am I wrong?

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (zzaimeche) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:578]: I think they could still upload a plaintext list of urls with invalid entries containing forbidden characters, so now this checks for that. Probably not a common problem but I figured it couldn't hurt to keep the check.

Comment (vzhestkov) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:578]: Do we have examples of repo URLs containing such characters?

**Comment by zzaimeche** [comment]
> In general looks like it tries to parse the mirrorlist as a plaintext even if it was detected that it's xml. Am I wrong?

I don't think it does, because of when it runs _replace_and_check_url() ; this is after it has tried to treat the list as xml and got a list of mirror urls from that (but it looks confusing because that method is defined between the try-excepts, but called later, and I might have said the wrong thing earlier...) I think the bigger problem is the way the surrounding code is nested/structured; the control flow isn't obvious and it makes things hard to follow.

(Earlier, I tried to double-check behaviour with an url pointing to a metalink and that failed, but at a later point in the process; the part I'd changed seemed to work. So I'm not sure if the xml parsing currently works on master...)

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (zzaimeche) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:578]: I've not seen any. (I based the list of forbidden characters on the response here: https://stackoverflow.com/a/36667242 . If we think there are people using urls that actually contain those characters somehow, I'll take it out again.)

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (vzhestkov) [inline python/spacewalk/satellite_tools/repo_plugins/yum_src.py:537]: maybe change it to somethinkg like:
```
if webpage.headers["Content-Type"] not in ("text/plain", "xml"...)
```
Just a tuple with the exact possible values for `Content-Type`.

**Review by vzhestkov — CHANGES_REQUESTED** [review]
In general it looks like this part of the code https://github.com/uyuni-project/uyuni/pull/6026/files#diff-a4f551c989af6141e94f0e9fcdfb392087152bed7218be989b28377c924812aaR597-R601 is used actually for both of `xml` and `text/plain` cases. Not sure if we really need to do it.

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (zzaimeche) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:537]: Oh, sorry, just checked the behaviour; I did it this way because the content_type is output in the form:
"text/plain; charset=UTF-8"
so it did need to be part of the output rather than the whole thing! I suppose it could split on the semicolon, but I think that would overcomplicate it, so unless there are times when it's going to include 'xml' or 'text/plain' but be the wrong type of document, I think it's better to leave it.

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (zzaimeche) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:540]: I'm not sure if the preferred course is to remove it entirely so it fails, or to handle each exception specifically; if it's the latter then I'd rather do that in a followup patch because it may take some time to work out what needs to be there.

Comment (meaksh) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:540]: It seems all exceptions raised by `requests` are inherited from `requests.exceptions.RequestException`, so we could use that one.

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (meaksh) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:578]: Hmmm, if we include this extra check, I would make some log/warning message to be displayed in case we are excluding an URL as invalid.

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (agraul) [inline python/spacewalk/satellite_tools/repo_plugins/yum_src.py:34]: Is it needed to include `requests`? I'm asking because we already use `urlgrabber` in reposync and, while I'd like to get rid of it, I think it's better to stick to one http library if possible.

Comment (zzaimeche) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:34]: Yeah, I was hoping to avoid importing anything but I couldn't find a tidy way to get the content-type headers with urlgrabber. It seemed like it was possible to _supply_ headers but not to retrieve them in a consistent format; I think you'd have to parse the page itself looking for something starting <!DOCTYPE> or <?xml> and then map it to types by hand, and that seemed fragile... So unless I missed something with urlgrabber I think requests is the less messy option.

File: python/spacewalk/satellite_tools/repo_plugins/yum_src.py
Comment (zzaimeche) [reply python/spacewalk/satellite_tools/repo_plugins/yum_src.py:578]: Good point, I'll add something for that.

File: python/spacewalk/spacewalk-backend.changes
Comment (meaksh) [inline python/spacewalk/spacewalk-backend.changes:4]: This is not needed, you just need to put your new entry on top of the existing ones:

```suggestion
```

**Review by meaksh — APPROVED** [review]
LGTM - thanks!

**Comment by meaksh** [comment]
@zzaimeche since there is a conflict here in the `spacewalk-backend.changes` file, you would need to rebase latest changes on "master" branch here to your branch before we can merge this.

**Comment by zzaimeche** [comment]
I'm going to dismiss Victor's review unless there are objections, since it wouldn't be fair to rerequest a review and I can't merge while changes are requested. He had suggested a change but the link to the relevant lines doesn't work for me now, so I can't find what it refers to, but it doesn't seem to be something I added. If there are no other objections I'll merge this.

**Pull request merged**
